plugins/omelasticsearch/omelasticsearch.c

```diff
@@ -247,76 +276,119 @@ ENDdbgPrintInstInfo
 
 
 /* Build basic URL part, which includes hostname and port as follows:
- * http://hostname:port/
- * Newly creates an estr for this purpose.
+ * http://hostname:port/ based on a server param
+ * Newly creates a cstr for this purpose.
  */
 static rsRetVal
-setBaseURL(instanceData *pData, es_str_t **url)
+computeBaseUrl(char* serverParam, int defaultPort, sbool useHttps, uchar **baseUrl)
 {
+#	define SCHEME_HTTPS "https://"
+#	define SCHEME_HTTP "http://"
+
 	char portBuf[64];
-	int r;
+	int r = 0;
+	char* host = serverParam;
 	DEFiRet;
 
-	*url = es_newStr(128);
-	snprintf(portBuf, sizeof(portBuf), "%d", pData->port);
-        if (pData->useHttps) {
-  		r = es_addBuf(url, "https://", sizeof("https://")-1);
+	es_str_t *urlBuf = es_newStr(256);
+	if (urlBuf == NULL) {
+		DBGPRINTF("omelasticsearch: failed to allocate es_str urlBuf in computeBaseUrl\n");
+		ABORT_FINALIZE(RS_RET_ERR);
+	}
+
+	/* Remove a trailing slash if it exists */
+	if (serverParam[strlen(serverParam)-1] == '/')
+		serverParam[strlen(serverParam)-1] = '\0';
+
+	/* Find where the hostname/ip of the server starts. If the scheme is not specified
+	  in the uri, start the buffer with a scheme corresponding to the useHttps parameter.
+	*/
+	if (strcasestr(serverParam, SCHEME_HTTP))
+		host = serverParam + strlen(SCHEME_HTTP);
+	else if (strcasestr(serverParam, SCHEME_HTTPS))
+		host = serverParam + strlen(SCHEME_HTTPS);
+	else
+		r = useHttps ?
+			es_addBuf(&urlBuf, SCHEME_HTTPS, sizeof(SCHEME_HTTPS)-1) : 
+			es_addBuf(&urlBuf, SCHEME_HTTP, sizeof(SCHEME_HTTP)-1);
+
+	if (r == 0) r = es_addBuf(&urlBuf, serverParam, strlen(serverParam));
+	if (r == 0 && !strchr(host, ':')) {
+		snprintf(portBuf, sizeof(portBuf), ":%d", defaultPort);
+		r = es_addBuf(&urlBuf, portBuf, strlen(portBuf));
+	}
+	if (r == 0) r = es_addChar(&urlBuf, '/');
+	if (r == 0) *baseUrl = (uchar*) es_str2cstr(urlBuf, NULL);
+
+	if (r != 0 || baseUrl == NULL) {
+		DBGPRINTF("omelasticsearch: error occurred computing baseUrl from server %s\n", serverParam);
+		ABORT_FINALIZE(RS_RET_ERR);
 	}
-	else {
-		r = es_addBuf(url, "http://", sizeof("http://")-1);
+finalize_it:
+	if (urlBuf) {
+		es_deleteStr(urlBuf);
 	}
-	if(r == 0) r = es_addBuf(url, (char*)pData->server, strlen((char*)pData->server));
-	if(r == 0) r = es_addChar(url, ':');
-	if(r == 0) r = es_addBuf(url, portBuf, strlen(portBuf));
-	if(r == 0) es_addChar(url, '/');
 	RETiRet;
 }
 
+static inline void
+incrementServerIndex(wrkrInstanceData_t *pWrkrData)
+{
+	pWrkrData->serverIndex = (pWrkrData->serverIndex + 1) % pWrkrData->pData->numServers;
+}
 
 static rsRetVal
 checkConn(wrkrInstanceData_t *pWrkrData)
 {
-	es_str_t *url;
-	CURL *curl = NULL;
+#	define HEALTH_URI "_cat/health"
```

Comment:
Testing _cat/health instead of just /, since if the server is up, but there is no master, '/' will respond successfully, but _cat/health will timeout indicating the cluster definitely cannot accept any messgaes. 

Not parsing the result of _cat/health, since if it can respond, we can defer to the cluster to decide whether or not it can accept messages